### ISSUE TITLE
Bias due to the input and output to the controller being both computed from marker trajectories

### ISSUE THREAD

**Issue by moorepants** [csu-hmc/gait-control-direct-id-paper#40]
The following graph shows example joint isolated gains identified from artificial data. I select the marker trajectories and ground reaction loads for a single gait cycle, fit a Fourier series to each measurement, repeat the gait cycles, add Gaussian noise to each measurement, and then run it through the data processing pipeline.

![gains-020](https://cloud.githubusercontent.com/assets/276007/7506851/bb50e3de-f41a-11e4-815e-c6df6634cb08.png)

This notebook shows the data processing methods:

http://nbviewer.ipython.org/github/csu-hmc/gait-control-direct-id-paper/blob/master/notebooks/marker_error_propogation.ipynb

In this case, it is obvious that something like the vertical GRF is present in the proportional gains. This makes sense, because if you want a model that converts angles and rates to joint torques you'd likely need something that  looks like ground reaction loads in your model. But the derivative gains seem like noise centered at zero, which is what you'd hope to get out of this test. I'm not sure what that means.

I'm not sure what do to about this. Should I some how subtract the results from this from the gains I identify from the real data?

This makes me think all we've really developed here is a way to compute the inverse dynamics with a statistical model instead of using multibody dynamics.

@tvdbogert Do you have any thoughts on this?

**Comment by moorepants** [comment]
Some thoughts on this from speaking with @monthubbard and Andy Ruina today. Andy and Mont both seemed to feel that taken at face value there is nothing wrong with deriving the "measurements" for the controller inputs and outputs from the same source data, i.e. it is not a problem that the joint angles and rates are generated from the marker coordinates and the joint torques are generated from the joint . Andy gave an example of trying to identify the control mechanism in a car's cruise control, saying that both the input and output to the system id problem could be the same position measurement of the vehicle (or acceleration pedal?). Actually I'm not sure I understood this.

But when I described the results shown here, i.e. that patterns emerge from the fabricated data, Andy said that maybe something is going on. He suggested that since the markers + random noise no longer satisfy the inverse model dynamics, that maybe we are introducing something into the joint torques.

Mont suggested making a simpler problem to try to demonstrate that there is an inherent correlation among the system state and the forces/torques generated from inverse dynamics. I may be able to derive a minimal example problem that has all of the characteristics of the walking control id problem we have but based on a 1 DOF inverted pendulum. He also mentioned the ape swinging problem that is similar to walking in that is in periodic, but with fewer DoFs. There may be some technical details that makes it not as simple as an idea as it seems, but I will give it a try.

At one point I asked if you have a 1 DoF mass spring damper system and directly measure it's position and an external force acting on it is it better to do system id with the measured force as the input and the measured position as the output or is it better to numerically differentiate the position to get a velocity "measurement" and then system id with two outputs. It isn't clear to me that increasing the number of measurements in this fashion would improve the uncertainty in the parameter estimates. I think it may make it worse due to the noise amplitude increase by the differentiation. Although this could depend if you weight the position or velocity fit more in the identification.

Another thing we discussed was that if the assume controller structure was of sufficient complexity, i.e. complex enough that it could represent the plant, then you could potentially identify the plant dynamics instead of the controller dynamics. This is related to the van der Kooik work but a little different. We know that the relationship between the joint angles/rates and the joint torques is the inverse dynamics model. If the controller structure is sufficiently complex enough to describe the inverse dynamics, i.e. constant proportional to omega dot, then maybe you'd accidentally identify plant dynamics. But our controller doesn't explicitly have any terms that are proportional to joint angular accelerations. So it isn't clear if this is even possible.

Another thing Mont felt strongly about was that we shouldn't numerically differentiate data to generate the measurements. Rather, we should have a sensor model (function of the state that generates the marker positions and ground reaction loads).

Bottom line is that we need to figure out what causes this. I had assumed that the fact that both the input and output "measurements" were derived at least partially from marker coordinates, that the identification would expose an inherent correlation that is imposed by the inverse dynamics.

**Comment by tvdbogert** [comment]
In slow movements such as walking, the joint moment is mostly statically related to joint angles: moment arm x GRF.  The moment arm is going to be somewhat proportional to the joint angles and this is why the GRF variations show up as a proportional gain in the controller identification after you add noise to the data.

This is more of a signal processing and instrumentation issue, rather than the "inverse plant" (van der Kooij) influence.  This comes from how we compute joint torques from ground reaction forces.

Fortunately, it seems that the derivative gains are not affected, and proportional gains from the real data seem substantially larger than this.

Can we estimate how much this could have affected the results from the actual data?  Maybe this: add random noise to the marker coordinates in the real data.  As the magnitude of this noise gets larger, the results from the real data will start looking more and more like the results from the fake data.  If it takes an unrealistic amount of noise to make the results look like the fake results, maybe we are safe.

We are not obligated to quantify all possible errors.  We can just discuss the possibility of certain errors.  On the other hand, I would want to have some confidence that the result we present has some validity (even if we do not present the sensitivity analysis on which that confidence would be based).

**Comment by moorepants** [comment]
> In slow movements such as walking, the joint moment is mostly statically related to joint angles: moment arm x GRF. The moment arm is going to be somewhat proportional to the joint angles and this is why the GRF variations show up as a proportional gain in the controller identification after you add noise to the data.

Thanks for that explanation. It sounds like it is reasonable, but it'd be nice to prove it. Seems if I divide the ankle joint torque by the joint angle I should get something similar in shape to the GRF, but that doesn't seem to be the case when I try, but that is of course ignoring the `m*(phi)` term in the actual calculation.

> This is more of a signal processing and instrumentation issue, rather than the "inverse plant" (van der Kooij) influence. This comes from how we compute joint torques from ground reaction forces.

I agree, but it is oddly similar. I was thinking about this loop scenario:

```
      -->[ Controller ]--
      |                 |
state |                 |  torque
      |                 |
      --[   Plant    ]<--
```

If you measure state and torque you can try to ID the controller of the plant: van der Kooij's problem. But if you only measure state, assume a plant, and estimate torque then try to id the controller you have van der Kooij's problem plus this potential to introduce these correlations from state to torque based on the plant which is similar to van der Kooij's problem but different. They almost seem like manifestations of the same thing.

> Fortunately, it seems that the derivative gains are not affected

Agreed.

> and proportional gains from the real data seem substantially larger than this.

Don't really here. Below I show the two plots to make the comparison: first is the mean result for 0.8 m/s for all subjects from real data and second is the result from a single 0.8 m/s trial from the artificial data plotted in the same mass normalized units.

From real data:
![mean-gains-0 8](https://cloud.githubusercontent.com/assets/276007/7600474/517a6e8a-f8c0-11e4-99f8-9fb949153907.png)

From fake data:
![gains-019](https://cloud.githubusercontent.com/assets/276007/7600366/4bb67706-f8bf-11e4-9bd8-3975f401d337.png)

Note that the proportional gains from each have comparable magnitudes. This is why I have concern.

> Can we estimate how much this could have affected the results from the actual data? Maybe this: add random noise to the marker coordinates in the real data. As the magnitude of this noise gets larger, the results from the real data will start looking more and more like the results from the fake data. If it takes an unrealistic amount of noise to make the results look like the fake results, maybe we are safe.

So I think you mean that I should do the same thing I do here except leave out the creation of the fabricate perfectly cyclic data, i.e. add increasing noise to the actual marker coordinates across all gait cycles. Then we can see what magnitude of noise causes the data to resemble the result presented here. Ok, I'll try that.

> We are not obligated to quantify all possible errors. We can just discuss the possibility of certain errors. On the other hand, I would want to have some confidence that the result we present has some validity (even if we do not present the sensitivity analysis on which that confidence would be based).

Sure, but this issue seems glaring to me. If all we are showing in the gain plots is this correlation due to our signal processing methods, then I have little confidence in stating any other result.

**Comment by moorepants** [comment]
Ok, I did the noise addition as you suggested. For the above plots generated from fake data I added 5 mm standard deviation Gaussian noise to all of the markers before the inverse dynamics. I now check 11 additive noise levels for the markers on the real data between 0 mm and 20 mm standard deviation. As the noise increases the graphs morph from the real data results to the fake data results shown above, as you suspected. Here are the gains for each noise level:

https://plus.google.com/photos/110966557175293116547/albums/6148485048844583425?authkey=CP6Ruobz57ahyQE

The 2mm noise doesn't show much change. By 4 mm you see some differences in gain magnitudes. And by 6 mm there is significant change. The 20mm looks a lot like the fake data results above.

So if the Motion Analysis specs are correct I think we are not likely to see anything close to 2 mm in random noise on marker measurements. What do you think?

**Comment by moorepants** [comment]
@monthubbard, Not sure if you care to read all this, but I'd be curious if you have any opinions on the recent conversation here since we spoke about this issue at length.

**Comment by monthubbard** [comment]
I appreciated you copying me to the conversation. I did read the messages, but I didn't follow it all. I still think pretty much what I did the other day: that if you are worried about the results not being meaningful then the simple model has a good shot at clarifying this.

I would like to read the van der kooij paper. Did you offer to send it to me?

Mont


On May 13, 2015, at 3:37 PM, Jason K. Moore <notifications@github.com<mailto:notifications@github.com>> wrote:


@monthubbard<https://github.com/monthubbard>, Not sure if you care to read all this, but I'd be curious if you have any opinions on the recent conversation here since we spoke about this issue at length.

**Comment by moorepants** [comment]
Mont,

Do you think this statement from Ton explains what I'm seeing?

> In slow movements such as walking, the joint moment is mostly statically related to joint angles: moment arm x GRF. The moment arm is going to be somewhat proportional to the joint angles and this is why the GRF variations show up as a proportional gain in the controller identification after you add noise to the data.
